statistics/handle/unlock_stats_handler.go

```diff
@@ -0,0 +1,126 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handle
+
+import (
+	"context"
+
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/statistics/handle/cache"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"go.uber.org/zap"
+)
+
+// RemoveLockedTables remove tables from table locked array.
+// - tids: table ids of which will be unlocked.
+// - pids: partition ids of which will be unlocked.
+// - tables: table names of which will be unlocked.
+// Return the message of skipped tables and error.
+func (h *Handle) RemoveLockedTables(tids []int64, pids []int64, tables []*ast.TableName) (string, error) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnStats)
+	exec := h.mu.ctx.(sqlexec.SQLExecutor)
+
+	_, err := exec.ExecuteInternal(ctx, "BEGIN PESSIMISTIC")
+	if err != nil {
+		return "", err
+	}
+
+	// Load tables to check locked before delete.
+	tableLocked, err := loadLockedTables(ctx, exec, maxChunkSize)
+	if err != nil {
+		return "", err
+	}
+	skippedTables := make([]string, 0, len(tables))
+
+	statsLogger.Info("unlock table", zap.Int64s("tableIDs", tids))
+	for i, tid := range tids {
+		var exist bool
+		exist, tableLocked = removeIfTableLocked(tableLocked, tid)
+		if !exist {
+			skippedTables = append(skippedTables, tables[i].Schema.L+"."+tables[i].Name.L)
+			continue
+		}
+		if err := updateStatsAndUnlockTable(ctx, exec, tid, maxChunkSize); err != nil {
+			return "", err
+		}
+	}
+
+	// Delete related partitions while don't warning delete empty partitions
+	for _, tid := range pids {
+		var exist bool
+		exist, tableLocked = removeIfTableLocked(tableLocked, tid)
+		if !exist {
+			continue
+		}
+		if err := updateStatsAndUnlockTable(ctx, exec, tid, maxChunkSize); err != nil {
+			return "", err
+		}
+	}
+
+	err = finishTransaction(ctx, exec, err)
+	if err != nil {
+		return "", err
+	}
+
+	// Update handle.tableLocked after transaction success, if txn failed, tableLocked won't be updated.
+	h.tableLocked = tableLocked
+
+	mag := generateSkippedTablesMessage(tids, skippedTables, unlockAction, unlockedStatus)
+	return mag, nil
+}
+
+func updateStatsAndUnlockTable(ctx context.Context, exec sqlexec.SQLExecutor, tid int64, maxChunkSize int) error {
+	count, modifyCount, version, err := getStatsDeltaFromTableLocked(ctx, tid, exec, maxChunkSize)
+	if err != nil {
+		return err
+	}
+
+	if _, err := exec.ExecuteInternal(ctx,
+		"UPDATE mysql.stats_meta SET version = %?, count = count + %?, modify_count = modify_count + %? WHERE table_id = %?",
+		version, count, modifyCount, tid); err != nil {
+		return err
+	}
+	cache.TableRowStatsCache.Invalidate(tid)
```

Comment:
I recommend removing this operation of refreshing Delta to make the whole unlocking process simpler since there is another background task goroutine updating all Delta data periodically and automatically.
In addition, a TiDB cluster has multiple TiDB instances, so just refreshing this instance's Delta seems useless, the Delta of this table is still inaccurate.
So I recommend removing this to make it simpler.

Reply:
> I recommend removing this operation of refreshing Delta to make the whole unlocking process simpler since there is another background task goroutine updating all Delta data periodically and automatically.

Good point!

Reply:
But in `dumpTableStatCountToKV` we also skip locked tables. This is the only chance to update the delta into the `mysql.stats_meta`.

Reply:
> In addition, a TiDB cluster has multiple TiDB instances, so just refreshing this instance's Delta seems useless, the Delta of this table is still inaccurate.

I didn't get this. Could you please explain it a bit?

Reply:
Each TiDB instance has its own `Delta Data`, and they use `Delta Data` to update some system tables periodically and asynchronously, so the effect of `Delta Data` on these system tables is "eventually consistent".
So triggering an update operation here doesn't change this attribute, it's useless and redundant.